src/handlers/v1/markets/[chainId]/[marketId]/zap-withdraw/index.ts

```diff
@@ -0,0 +1,198 @@
+  handler: async (event) => {
+    const chainId = Number(event.pathParameters.chainId);
+    const marketId = event.pathParameters.marketId.toUpperCase();
+    if (!validateMarket(chainId, marketId)) {
+      throw newHttpError(400, { code: '400.1', message: 'market does not exist' });
+    }
+
+    if (!event.body) {
+      throw newHttpError(400, { code: '400.2', message: 'body is invalid' });
+    }
+
+    let account = event.body.account;
+    if (!account) {
+      throw newHttpError(400, { code: '400.3', message: `account can't be blank` });
+    }
+    try {
+      account = utils.getAddress(account);
+    } catch {
+      throw newHttpError(400, { code: '400.4', message: 'account is invalid' });
+    }
+
+    const service = new Service(chainId);
+
+    let marketInfo: MarketInfo;
+    try {
+      marketInfo = await service.getMarketInfo(marketId, account);
+    } catch (err) {
+      throw newInternalServerError(err);
+    }
+    const { utilization, healthRate, liquidationThreshold, borrowUSD, collateralUSD, netAPR } = marketInfo;
+    const currentPosition = { utilization, healthRate, liquidationThreshold, borrowUSD, collateralUSD, netAPR };
+
+    let targetTokenAmount = '0';
+    const logics: GetZapWithdrawQuotationResponseBody['logics'] = [];
+    let approvals: GetZapWithdrawQuotationResponseBody['approvals'] = [];
+    let permitData: GetZapWithdrawQuotationResponseBody['permitData'];
+    let targetPosition = currentPosition;
+    if (event.body.withdrawalToken && event.body.amount && event.body.targetToken && Number(event.body.amount) > 0) {
+      const { amount, slippage } = event.body;
+      const {
+        baseToken,
+        supplyAPR,
+        supplyBalance,
+        supplyUSD,
+        borrowAPR,
+        borrowCapacityUSD,
+        liquidationLimit,
+        baseTokenPrice,
+        collaterals,
+      } = marketInfo;
+
+      const withdrawalToken = common.Token.from(event.body.withdrawalToken);
+      const targetToken = common.Token.from(event.body.targetToken);
+
+      // 1. check withdraw collateral or base token
+      const withdrawalCollateral = collaterals.find(({ asset }) => asset.is(withdrawalToken.unwrapped));
+      if (!withdrawalCollateral && !withdrawalToken.unwrapped.is(baseToken)) {
+        throw newHttpError(400, { code: '400.5', message: 'withdrawal token is not collateral nor base' });
+      }
+
+      // 2. new and append compound v3 withdraw logic
+      let realWithdrawalToken: common.Token;
+      if (withdrawalToken.unwrapped.is(baseToken)) {
+        if (new BigNumberJS(supplyBalance).lt(new BigNumberJS(amount))) {
+          throw newHttpError(400, {
+            code: '400.6',
+            message: 'withdrawal amount is greater than available base amount',
+          });
+        }
+        const cToken = await service.getCToken(marketId);
+        realWithdrawalToken = targetToken.wrapped.is(baseToken.wrapped) ? targetToken : withdrawalToken.wrapped;
```

Comment:
I'm a little confused. `targetToken` is the token being zapped out right? Why does it involve in step 2, which should be the withdrawal logic of Compound.

Reply:
The idea here is to use the `wrapMode` in Protocolink. If `targetToken` is native token, we can do the swap in withdrawal logic instead of doing the swap in Paraswap or other gateways.

Reply:
I see!